woocommerce/woocommerce
# WIP: Added new tests to cover editing product variations in the Product Edit page.

**Update**: this PR is still work in progress, I need to fix something.

### All Submissions:

* [ x ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ x ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ x ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Added the following tests to `product-edit.test.js`:
- Can edit just a single attribute of a product variation.
- Can edit multiple attributes of a product variation.

Initially, we could think that adding `can edit multiple attributes of a product variation` might be enough, as it also covers the other test. But based on [this issue](https://github.com/woocommerce/woocommerce/issues/31347), it's possible that the issue described there is only covered when only one attribute is updated and the cursor is not placed on any other attribute before clicking the `Save changes` button. Therefore, I think it would be a good idea adding these two tests, but this is open to discussion, of course.

Apart from these tests, the following functions were added to `merchant.js`:
- **updateVariationDetails**: update the details of a product variation.
- **verifyVariationDetails**: verify the details of a product variation.

Finally, the following functions were added to `page-utils.js`:
- **productPageSaveChanges**: save changes when editing a product. We already have one for orders, so I thought it would make sense to have one for products.
- **click**: asserts that the element to be clicked is enabled, and click it afterwards. I found it necessary because Puppeteer considers valid clicking an element that is disabled, and this might cause false positives. With this new function, if the element (for example, a button), is disabled, then the test will fail because it won't be able to proceed with the expected flow.

Regarding the `click` function, I think it would be good to have a module that contained low-level functions on top of our webdriver or UI automation library so that we could override these basic functions and make them more robust. Of course, it doesn't make sense to do it right now with Puppeteer, but I'd like to raise this discussion thinking about Playwright, @woocommerce/solaris and specially @lanej0. If Playwright is not quite robust out of the box (as most of these kinds of libraries, like Puppeteer, Selenium, etc.), do you think it would make sense adding this module to enforce Playwright interactions with elements?

### How to test the changes in this Pull Request:

1. Go to woocommerce/plugins/woocommerce.
2. Run the order-edit tests: `pnpx wc-e2e test:e2e ./tests/e2e/specs/wp-admin/product-edit.test.js`.

Alternative (this will run all the tests instead of just those related to editing the product information):
1. Go to the main woocommerce repo folder.
2. Run the e2e tests: `pnpm nx test-e2e woocommerce`.

**Note**: the new tests will keep failing until the fix for the [issue](https://github.com/woocommerce/woocommerce/issues/31347) is in place, which - as far as I know - is currently under development. I could use the `Puppeteer debugger` to modify the actual flow and verify the behaviour of these tests, but we won't be able to get a successful test run until the fix is ready.

### Other information:

* [ x ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ? ] Have you written new tests for your changes, as applicable?  
-> As far as I know, this is not applicable. Do we have unit tests for our e2e test project?
* [ x ] Have you successfully run tests with your changes locally?

### Changelog entry

See `Changes proposed in this Pull Request`.

### FOR PR REVIEWER ONLY:
